### ISSUE TITLE
my theme already has katex shortcode

### ISSUE THREAD

**Issue by longlu** [kaushalmodi/ox-hugo#377]
the theme I use already have shortcode for katex and mermaid etc.
the shortcode for the markdown file is like this:
``{{< katex display >}}``
``E = -J \sum\_{i=1}^N s\_i s\_{i+1}``
``{{< /katex >}}``

my settings in org file is this:
** article 1
``:PROPERTIES:``
``:EXPORT_FILE_NAME: xxxx``
``:EXPORT_HUGO_URL: xxxx``
``:EXPORT_HUGO_PAIRED_SHORTCODES: <katex katex <mermaid mermaid <tab tab``
:END:
*** test heading 1
``\( E = -J \sum_{i=1}^N s_i s_{i+1} \)``

``#+attr_shortcode: display``
``#+begin_katex``
``E = -J \sum_{i=1}^N s_i s_{i+1}``
``#+end_katex``

but after exporting,  the formula got converted to this in the exported md file:
``{{< katex display >}}``
``E = -J &sum;<sub>i=1</sub>^N s\_i s<sub>i+1</sub>``
``{{< /katex >}}``

I think conversion like this is ok for some type of shortcode, but in this case the formula does not need to be changed, 
how can I export a correct shortcode for this?

Another question is, how can i generate non paired shortcode:

for example:   if i want this : ``{{< section >}}`` in my exported markdown file, what should i add in my org file?

**Comment by kaushalmodi** [comment]
Hello, 

As I designed ox-hugo to have first-class Org support, the Hugo shortcode support is only minimal. I would suggest using Org macros where possible instead. That allows the same Org document to be exportable in other formats too.

About the katex shortcode:

Org treats the content in Org special blocks (like the `#+begin_katex ..`) as normal content and so the same escaping rules apply there too.

I have not used katex. But can you use the Org equations support as shown in https://ox-hugo.scripter.co/doc/equations/ and then have the katex scripts take care of those? Do you need the katex shortcode here?


>how can i generate non paired shortcode:

You can do it this way: https://github.com/kaushalmodi/ox-hugo/issues/126#issuecomment-444914066

---

Finally, it you really really need to use the katex shortcode, you can do:

```org
#+begin_export html
{{< katex display >}}
E = -J \sum\_{i=1}^N s\_i s\_{i+1}
{{< /katex >}}
#+end_export
```

Ref: https://github.com/kaushalmodi/ox-hugo/issues/126#issue-293883528

**Comment by longlu** [comment]
>  But can you use the Org equations support as shown in https://ox-hugo.scripter.co/doc/equations/ and then have the katex scripts take care of those? Do you need the katex shortcode here?

I believe it can be done that way, i don't know how to do it tho, as i'm not the theme developer.  (https://themes.gohugo.io/theme/hugo-book/docs/shortcodes/katex/)

I tested the equations sections before, it also works for inline equations, but i'm not sure wheths it's displayed by mathjax or katex.


This doesn't display as correctly tho.
``\begin{equation}``
``\label{eq:1}``
``C = W\log_{2} (1+\mathrm{SNR})``
``\end{equation}``

**Comment by kaushalmodi** [comment]
Hello, sorry for the late reply. I have documented the solution for your initial question regarding `katex` shortcode in https://github.com/kaushalmodi/ox-hugo/commit/e2bc6c4f8dceb0c483ee6456df70a8cf5f5c2e48 (it's possible to export that using `#+begin_export hugo` as I mentioned earlier.

But as for your followup comment:

> This one doesn't display correctly ..

I need more context.

Do you have a snippet of that equation from Markdown file that works correctly? If you paste that here, I can help you with how to write the equivalent in Org.

**Comment by kaushalmodi** [comment]
I'll close this issue until I hear more. Feel free to reply in this thread even after the issue is closed.

**Comment by longlu** [comment]
thanks , I'm still learning

$C = W\log_{2} (1+\mathrm{SNR})$

it can be correctly displayed with this^

**Comment by kaushalmodi** [comment]
It displays fine with MathJax:

![image](https://user-images.githubusercontent.com/3578197/148174810-a7b299d7-3116-4412-b4f2-29a8bca09aa3.png)

I don't have a theme setup with KaTeX. But if you type something in Markdown and that works well with your KaTeX theme site, share your Markdown file and I can help you get the Org version.

**Comment by kaushalmodi** [comment]
I have a bunch of tests and examples for equations. Please check out https://ox-hugo.scripter.co/test/tags/equations. When you visit a page, you will find links to the Markdown source generated by ox-hugo. The Org sources for all tests are in this repo's `test/site/content-org/` directory,

**Comment by longlu** [comment]
https://hugo-book-demo.netlify.app/docs/shortcodes/katex/   <--- this is a katex supported theme.


$C = W\log_{2} (1+\mathrm{SNR})$     < --this line display the fomulat correctly 

this also works:

#+begin_export hugo
{{< katex display >}}
E = -J \sum\_{i=1}^N s\_i s\_{i+1}
{{< /katex >}}
#+end_export





but this doesn't work:

\begin{equation}
\label{eq:1}
C = W\log_{2} (1+\mathrm{SNR})
\end{equation}









this was copied from here:   https://ox-hugo.scripter.co/test/posts/indented-equations/

it does not display formula with this 4 lines(copied from your examples):  

\begin{equation}
\label{eq:1}
C = W\log_{2} (1+\mathrm{SNR})
\end{equation}

**Comment by longlu** [comment]
\( E = -J \sum_{i=1}^N s_i s_{i+1} \)   <--- this also works with that theme.   ( although I don't know if used katex or mathjax to display.


I'm not sure why this does not work:

\begin{equation}
\label{eq:1}
C = W\log_{2} (1+\mathrm{SNR})
\end{equation}

**Comment by kaushalmodi** [comment]
> although I don't know if used katex or mathjax to display

Look at the page source and search for ".js". See what script is sourced. If your theme has a katex shortcode, then it's most likely using katex JavaScript code. 

I have tested all equations using only Mathjax. So it's possible that certain syntax might not work with katex.

**Comment by kaushalmodi** [comment]
https://github.com/KaTeX/KaTeX/issues/445

Ok, it looks like a katex bug. Please switch to a theme that uses Mathjax or add Mathjax script in header by modifying your theme. 

You can use the test site's theme as reference: https://github.com/kaushalmodi/hugo-bare-min-theme (it's just one small script in static/ and 2 lines in mathjax.html partial).

**Comment by kaushalmodi** [comment]
I'll close this issue as this is not something ox-hugo can fix. But I'd can discuss more on https://github.com/kaushalmodi/ox-hugo/discussions. 

If you open a discussion issue there, it could have better visibility for other users facing the same issue as yours.

**Comment by kaushalmodi** [comment]
@longlu I have just added a commit that will now allow you to use:

```org
#+attr_shortcode: display
#+begin_katex
E = -J \sum_{i=1}^N s_i s_{i+1}
#+end_katex
```

**Comment by longlu** [comment]
thanks

it seems i need to add that to the template decribed here:   https://orgmode.org/manual/Structure-Templates.html

that's the only shortcut I'm aware of

**Comment by longlu** [comment]
"#+attr_shortcode: display"

this one can not be automatically inserted with that shortcut

**Comment by kaushalmodi** [comment]
> that's the only shortcut I'm aware of

Emacs has a lot of ways for inserting snippets. Please look into yasnippet and similar packages. 

Alternatively, you can even tweak that Hugo shortcode so that the "display" option behavior is enabled by default.
